File: src/main/java/com/mynimef/bot/config/CallbacksBuilder.java
package com.mynimef.bot.config;

import com.mynimef.bot.actions.ICallback;

import java.util.HashMap;
import java.util.Map;

public abstract class CallbacksBuilder {
    private final Map<String, ICallback> callbacks;

    public CallbacksBuilder() {
        this.callbacks = new HashMap<>();
    }

    protected abstract void init();

    public Map<String, ICallback> getCallbacks() {
        init();
        return callbacks;
    }

    protected void add(String callbackId, ICallback callback) {
        callbacks.put(callbackId, callback);
    }
}

File: src/main/kotlin/com/mynimef/bot/config/CallbacksBuilder.kt
package com.mynimef.bot.config

import com.mynimef.bot.executable.ActionCallback
import okhttp3.internal.toImmutableMap


abstract class CallbacksBuilder {

    private val _callbacks: MutableMap<String, ActionCallback> = HashMap()
    internal val callbacks by lazy {
        init()
        _callbacks.toImmutableMap()
    }

    protected abstract fun init()

    protected fun add(callbackId: String, action: ActionCallback) {
        _callbacks[callbackId] = action
    }

}
